### ISSUE TITLE
ChildRules with class hierarchies broken

### ISSUE THREAD

**Issue by runehalfdan** [FluentValidation/FluentValidation#2165]
### FluentValidation version

11.7.1

### Summary

#2097 (Fix for multi-level child-rules) broke our validation code. We get exception `Unable to cast object of type FluentValidation.Internal.RuleBuilder2[Root, Bar] to type FluentValidation.Internal.RuleBuilder2[Root, Base]` when setting up validation in sample below. 

It worked with 11.5.1, but broke when we upgraded to 11.7.1

### Steps to Reproduce

```
using FluentValidation;

public class Root
{
    public Bar Data {get; set;} = null;
}

public class Base 
{
    public int Value {get; set;}
}

public class Bar : Base
{
    public int BarValue {get; set;}
}

public class RootValidator : AbstractionValidator<Root>
{
    public RootValidator()
    {
        RuleFor(x => x).ChildRules(RootRules);
    }

    public static Action<InlineValidator<Base>> BaseRules()
    {
        return rules => {
            rules.RuleFor(x => x.Value).NotEqual(-1);
        };
    }

    public static Action<InlineValidator<Root>> RootRules()
    {
        return rules => {
            rules.RuleFor(x => x.Data).ChildRules(BaseRules());
        };
    }
}

public static class Program
{
    public static void Main()
    {
        var validator = new RootValidator();
        var result = validator.Valdiate(new Root());
    }
}
```

**Comment by runehalfdan** [comment]
We have worked around it for now, by being more explicit about types in the various rule-builders (either just using the most specific type or using constrained generics). But is very troublesome that this causes runtime exceptions instead compile-time errors.

**Comment by JeremySkinner** [comment]
Fixed in a437761052237fdf1102b55014ba6ed1da0f2411 which will be part of the 11.8 release.

**Comment by JeremySkinner** [comment]
11.8 is released with this change
